### PR TITLE
Remove Hydrogen dependence

### PR DESCRIPTION
**Pull request by benson31**
The "big things" that happened in the C++ code here are:

1. Migration of @tbennun's version of the GPU caching allocator from Hydrogen.
2. Explicit dependence on Aluminum.
3. Quick-and-dirty `Comm` wrapper to meet the current (very minimal) communication requirements in H2.

The rest was fairly boiler-plate.

[CI is happy.](https://lc.llnl.gov/gitlab/benson31/h2/-/pipelines/469265)

### CODE REVIEW

**Review by ndryden — APPROVED** [review]
That's a smoother extraction than I thought it would be. Looks good to me.

File: test/unit_test/mpi_event_listener.cpp [hunk 1]
```diff
@@ -55,11 +57,11 @@ class MPIEventListener final : public Catch::EventListenerBase
...
     {
       int test_result = 0;
-      El::mpi::AllReduce(&test_result,
-                         1,
-                         El::mpi::MIN,
-                         El::mpi::COMM_WORLD,
-                         El::SyncInfo<El::Device::CPU>{});
+      h2_tmp::allreduce(&test_result,
+                        1,
+                        Al::ReductionOperator::land,
```
Comment (tbennun) [inline test/unit_test/mpi_event_listener.cpp:62]: so the minimum changed to a logical and? If it's safe to do with integers (and it probably is), I'm fine with it (shouldn't it be an arithmetic and?)

**Review by tbennun — APPROVED** [review]
Wow, the dependency was removed and the code is cleaner for it.

File: test/unit_test/mpi_event_listener.cpp (hunk 1 above)
Comment (benson31) [reply test/unit_test/mpi_event_listener.cpp:62]: Yeah, this is safe. There's even an example with it in the MPI spec (Example 11.13 of the 4.1 spec) with `MPI_INT` as the datatype. Re "arithmetic and", the spec gives you two options: `MPI_LAND` and `MPI_BAND` (aw shucks, I've always wanted to be in a band...), so I'm sticking with the logical :)

**Pull request merged**